AceComm-3.0/ChatThrottleLib.lua

```diff
@@ -292,38 +314,89 @@ end
 -- - and each pipe contains messages
 
+local SendAddonMessageResult = Enum.SendAddonMessageResult or {
+	Success = 0,
+	AddonMessageThrottle = 3,
+	NotInGroup = 5,
+	ChannelThrottle = 8,
+	GeneralError = 9,
+}
+
+local function MapToSendResult(result)
+	if result == true then
+		result = SendAddonMessageResult.Success
+	elseif result == false then
+		result = SendAddonMessageResult.GeneralError
+	end
+
+	return result
+end
+
+local function IsThrottledSendResult(result)
+	return result == SendAddonMessageResult.AddonMessageThrottle
+end
+
+local function PerformSend(sendFunction, ...)
+	-- The select(-1, true, ...) incantation grabs the last return value
+	-- from the called function, defaulting to true if it returned nothing.
+	--
+	-- The default case is required for SendChatMessage. The selection of
+	-- the last return value is required due to per-flavor API divergences.
+
+	bMyTraffic = true
+	local sendResult = select(-1, true, sendFunction(...))
+	bMyTraffic = false
+
+	return MapToSendResult(sendResult)
+end
+
 function ChatThrottleLib:Despool(Prio)
 	local ring = Prio.Ring
 	while ring.pos and Prio.avail > ring.pos[1].nSize do
-		local msg = table_remove(ring.pos, 1)
-		if not ring.pos[1] then  -- did we remove last msg in this pipe?
-			local pipe = Prio.Ring.pos
-			Prio.Ring:Remove(pipe)
-			Prio.ByName[pipe.name] = nil
-			DelPipe(pipe)
-		else
-			Prio.Ring.pos = Prio.Ring.pos.next
-		end
-		local didSend=false
+		local pipe = ring.pos
+		local msg = pipe[1]
+		local sendResult
 		local lowerDest = strlower(msg[3] or "")
+
 		if lowerDest == "raid" and not UnitInRaid("player") then
 			-- do nothing
+			sendResult = SendAddonMessageResult.NotInGroup
 		elseif lowerDest == "party" and not UnitInParty("player") then
 			-- do nothing
+			sendResult = SendAddonMessageResult.NotInGroup
 		else
-			Prio.avail = Prio.avail - msg.nSize
-			bMyTraffic = true
-			msg.f(unpack(msg, 1, msg.n))
-			bMyTraffic = false
-			Prio.nTotalSent = Prio.nTotalSent + msg.nSize
-			DelMsg(msg)
-			didSend = true
+			sendResult = PerformSend(msg.f, unpack(msg, 1, msg.n))
 		end
-		-- notify caller of delivery (even if we didn't send it)
-		if msg.callbackFn then
-			msg.callbackFn (msg.callbackArg, didSend)
+
+		if IsThrottledSendResult(sendResult) then
+			-- Message was throttled; move the pipe into the blocked ring.
+			Prio.Ring:Remove(pipe)
+			Prio.Blocked:Add(pipe)
+		else
+			-- Dequeue message after submission.
+			table_remove(pipe, 1)
+			DelMsg(msg)
+
+			if not pipe[1] then  -- did we remove last msg in this pipe?
+				Prio.Ring:Remove(pipe)
+				Prio.ByName[pipe.name] = nil
+				DelPipe(pipe)
+			else
+				ring.pos = ring.pos.next
+			end
+
+			-- Update bandwidth counters on successful sends.
+			local didSend = (sendResult == SendAddonMessageResult.Success)
+			if didSend then
+				Prio.avail = Prio.avail - msg.nSize
+				Prio.nTotalSent = Prio.nTotalSent + msg.nSize
+			end
+
+			-- Notify caller of delivery.
```

Comment:
Comment is misleading as its not really delivery but a send

Reply:
Original comment, just minus a few words - but it makes sense to change it.